src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -0,0 +1,696 @@
+    return {.in = lhs.in - rhs.in, .eg = lhs.eg - rhs.eg};
+}
+template<std::integral T>
+ingress_egress_state<T> operator-(const ingress_egress_state<T>& v) {
+    return {.in = -v.in, .eg = -v.eg};
+}
+template<std::integral T>
+bool is_zero(const ingress_egress_state<T>& v) noexcept {
+    return v.in == 0 && v.eg == 0;
+}
+
+/// Copies vector of inoutpairs to an ingress_egress_state of vectors
+template<class T>
+ingress_egress_state<std::vector<T>>
+to_soa_layout(const std::vector<ingress_egress_state<T>>& v) {
+    ingress_egress_state<std::vector<T>> res;
+    res.in.reserve(v.size());
+    res.eg.reserve(v.size());
+    for (const ingress_egress_state<T>& i : v) {
+        res.in.push_back(i.in);
+        res.eg.push_back(i.eg);
+    };
+    return res;
+}
+
+namespace {
+
+quota_t node_to_shard_quota(const std::optional<quota_t> node_quota) {
+    if (node_quota.has_value()) {
+        const quota_t v = node_quota.value() / ss::smp::count;
+        if (v > bottomless_token_bucket::max_quota) {
+            return bottomless_token_bucket::max_quota;
+        }
+        if (v < 1) {
+            return quota_t{1};
+        }
+        return v;
+    } else {
+        return bottomless_token_bucket::max_quota;
+    }
+}
+
+} // namespace
+
+snc_quota_manager::snc_quota_manager()
+  : _max_kafka_throttle_delay(
+    config::shard_local_cfg().max_kafka_throttle_delay_ms.bind())
+  , _kafka_throughput_limit_node_bps{
+      config::shard_local_cfg().kafka_throughput_limit_node_in_bps.bind(),
+      config::shard_local_cfg().kafka_throughput_limit_node_out_bps.bind()}
+  , _kafka_quota_balancer_window(
+      config::shard_local_cfg().kafka_quota_balancer_window.bind())
+  , _kafka_quota_balancer_node_period(
+      config::shard_local_cfg().kafka_quota_balancer_node_period.bind())
+  , _kafka_quota_balancer_min_shard_thoughput_ratio(
+      config::shard_local_cfg()
+        .kafka_quota_balancer_min_shard_thoughput_ratio.bind())
+  , _kafka_quota_balancer_min_shard_thoughput_bps(
+      config::shard_local_cfg()
+        .kafka_quota_balancer_min_shard_thoughput_bps.bind())
+  , _node_quota_default{calc_node_quota_default()}
+  , _shard_quota{
+      .in {node_to_shard_quota(_node_quota_default.in),
+           _kafka_quota_balancer_window()},
+      .eg {node_to_shard_quota(_node_quota_default.eg),
+           _kafka_quota_balancer_window()}}
+{
+    update_shard_quota_minimum();
+    _kafka_throughput_limit_node_bps.in.watch(
+      [this] { update_node_quota_default(); });
+    _kafka_throughput_limit_node_bps.eg.watch(
+      [this] { update_node_quota_default(); });
+    _kafka_quota_balancer_window.watch([this] {
+        const auto v = _kafka_quota_balancer_window();
+        vlog(klog.debug, "qm - Set shard TP token bucket window: {}", v);
+        _shard_quota.in.set_width(v);
+        _shard_quota.eg.set_width(v);
+    });
+    _kafka_quota_balancer_node_period.watch([this] {
+        if (_balancer_timer.cancel()) {
```

Comment:
`maybe_arm_balance_timer` checks that the `gate` is open, however its possible this could be called before the gate has opened